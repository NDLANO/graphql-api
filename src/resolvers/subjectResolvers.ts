/**
 * Copyright (c) 2019-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

// @ts-strict-ignore

import { ISubjectPageData } from '@ndla/types-backend/frontpage-api';
import { Node } from '@ndla/types-taxonomy';
import { fetchLK20CompetenceGoalSet } from '../api';

import {
  GQLQuerySubjectArgs,
  GQLSubject,
  GQLSubjectLink,
  GQLTopic,
} from '../types/schema';
import { filterMissingArticles } from '../utils/articleHelpers';

export const Query = {
  async subject(
    _: any,
    { id }: GQLQuerySubjectArgs,
    context: ContextWithLoaders,
  ): Promise<Node> {
    return await context.loaders.subjectLoader.load({ id });
  },
  async subjects(
    _: any,
    input:
      | {
          metadataFilterKey?: string;
          metadataFilterValue?: string;
          filterVisible?: boolean;
        }
      | undefined,
    context: ContextWithLoaders,
  ): Promise<GQLSubject[]> {
    const metaDataFilter = input?.metadataFilterKey
      ? {
          metadataFilter: {
            key: input.metadataFilterKey,
            value: input.metadataFilterValue,
          },
        }
      : {};

    const loaderParams = {
      ...metaDataFilter,
      filterVisible: input.filterVisible,
    };

    return context.loaders.subjectsLoader
      .load(loaderParams)
      .then(s => s.subjects);
  },
};

export const resolvers = {
  Subject: {
    async topics(
      subject: GQLSubject,
      args: { all: boolean },
      context: ContextWithLoaders,
    ): Promise<GQLTopic[]> {
      const topics = await context.loaders.subjectTopicsLoader.load({
        subjectId: subject.id,
      });
      if (args.all) {
        return filterMissingArticles(topics, context);
      }
      return filterMissingArticles(
        topics.filter((topic: GQLTopic) => topic.parent === subject.id),
        context,
      );
    },
    async allTopics(subject: GQLSubject, _: any, context: ContextWithLoaders) {
      const topics = await context.loaders.subjectTopicsLoader.load({
        subjectId: subject.id,
      });
      return filterMissingArticles(topics, context);
    },
    async subjectpage(
      subject: GQLSubject,
      __: any,
      context: ContextWithLoaders,
    ): Promise<ISubjectPageData | undefined> {
      if (subject.contentUri?.startsWith('urn:frontpage')) {
        return context.loaders.subjectpageLoader.load(
          subject.contentUri.replace('urn:frontpage:', ''),
        );
      }
    },
    async grepCodes(
      subject: GQLSubject,
      __: any,
      context: ContextWithLoaders,
    ): Promise<string[]> {
      if (subject.metadata?.grepCodes) {
        const code = subject.metadata?.grepCodes?.find(c => c.startsWith('KV'));
        return code ? fetchLK20CompetenceGoalSet(code, context) : [];
      }
      return [];
    },
  },
  SubjectPage: {
    async connectedTo(
      subjectpage: ISubjectPageData,
      _: any,
      context: ContextWithLoaders,
    ): Promise<GQLSubjectLink[]> {
      return await context.loaders.subjectLoader.loadMany(
        subjectpage.connectedTo.map(id => {
          return { id };
        }),
      );
    },
    async buildsOn(
      subjectpage: ISubjectPageData,
      _: any,
      context: ContextWithLoaders,
    ): Promise<GQLSubjectLink[]> {
      return await context.loaders.subjectLoader.loadMany(
        subjectpage.buildsOn.map(id => {
          return { id };
        }),
      );
    },
    async leadsTo(
      subjectpage: ISubjectPageData,
      _: any,
      context: ContextWithLoaders,
    ): Promise<GQLSubjectLink[]> {
      return await context.loaders.subjectLoader.loadMany(
        subjectpage.leadsTo.map(id => {
          return { id };
        }),
      );
    },
  },
};
