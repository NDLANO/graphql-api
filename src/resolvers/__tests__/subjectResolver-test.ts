/**
 * Copyright (c) 2020-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

const nock = require('nock');
import DataLoader from 'dataloader';
import { Query } from '../subjectResolvers';

test('Fetch subject should filter out invisible elements', async () => {
  const subjects = [
    {
      id: 'urn:subject:3',
      contentUri: 'urn:frontpage:1',
      name: 'Samfunnsfag YF Vg2',
      path: '/subject:3',
      metadata: {
        visible: true,
      },
    },
    {
      id: 'urn:subject:6',
      contentUri: 'urn:frontpage:4',
      name: 'BrÃ¸nnteknikk Vg2',
      path: '/subject:6',
      metadata: {
        visible: false,
      },
    },
  ];
  nock('https://api.test.ndla.no')
    .persist()
    .get('/taxonomy/v1/subjects/?language=nb')
    .reply(200, subjects);

  const loadAll = async () => {
    return [{ subjects }];
  };

  const subjectsLoader = new DataLoader(loadAll);

  const subs = await Query.subjects(1, 1, {
    language: 'nb',
    shouldUseCache: false,
    taxonomyUrl: 'taxonomy',
    loaders: { subjectsLoader },
  });

  expect(subs).toMatchSnapshot();
});

test('Fetch subject filters should filter out invisible elements', async () => {
  nock('https://api.test.ndla.no')
    .persist()
    .get('/taxonomy/v1/filters/?language=nb')
    .reply(200, [
      {
        id: 'urn:filter:1',
        name: 'Filter 1',
        metadata: {
          grepCodes: [],
          visible: true,
        },
      },
      {
        id: 'urn:filter:2',
        name: 'Filter 2',
        metadata: {
          grepCodes: [],
          visible: false,
        },
      },
    ]);

  const subs = await Query.filters(1, 1, {
    language: 'nb',
    shouldUseCache: false,
    taxonomyUrl: 'taxonomy',
  });

  expect(subs).toMatchSnapshot();
});
