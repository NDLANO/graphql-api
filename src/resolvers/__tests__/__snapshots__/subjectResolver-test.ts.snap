// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Fetch subject should filter out invisible elements 1`] = `
Array [
  Object {
    "contentUri": "urn:frontpage:1",
    "id": "urn:subject:3",
    "metadata": Object {
      "visible": true,
    },
    "name": "Samfunnsfag YF Vg2",
    "path": "/subject:3",
  },
]
`;
